src/Microsoft.PowerShell.Commands.Management/commands/management/TestConnectionCommand.cs

```diff
@@ -762,31 +790,282 @@ public void Dispose()
         /// </param>
         protected virtual void Dispose(bool disposing)
         {
-            if (!this._disposed)
+            if (!_disposed)
             {
                 if (disposing)
                 {
-                    _sender.Dispose();
+                    _sender?.Dispose();
+                    _pingComplete?.Dispose();
                 }
 
                 _disposed = true;
             }
         }
 
-        // Count of pings sent per each trace route hop.
-        // Default = 3 (from Windows).
-        // If we change 'DefaultTraceRoutePingCount' we should change 'ConsoleTraceRouteReply' resource string.
-        private const int DefaultTraceRoutePingCount = 3;
+        // Uses the SendAsync() method to send pings, so that Ctrl+C can halt the request early if needed.
+        private PingReply SendCancellablePing(
+            IPAddress targetAddress,
+            int timeout,
+            byte[] buffer,
+            PingOptions pingOptions,
+            Stopwatch timer = null)
+        {
+            try
+            {
+                _sender = new Ping();
+                _sender.PingCompleted += OnPingComplete;
 
-        /// Create the default send buffer once and cache it.
-        private const int DefaultSendBufferSize = 32;
-        private static byte[] s_DefaultSendBuffer = null;
+                timer?.Start();
+                _sender.SendAsync(targetAddress, timeout, buffer, pingOptions, this);
```

Comment:
Basically this allows us to accept and respect Ctrl+C during a ping request rather than having to wait for the timeout if no response is being received.

Reply:
sender.Send() has timeout and Ctrl-C also works before the change.

Reply:
In my testing, Ctrl+C has only been recognised after a PingReply has been received. Has something changed in .NET Core recently? I am aware .Send() has a timeout, but there is no capability to cancel while waiting for a reply; it is a blocking call.

Reply:
It is short timeout. Even for interactive session. Also we can decrease it from 5 to 2 seconds - modern channels is fast and high quality (no drop packets), it is not 36.8 kbit/s modems. And user can set custom timeout if you want.
I vote for a simple code and default 2 sec timeout.

Reply:
I think this is needed precisely because the user _can_ select a custom timeout value. If the user enters a long timeout, they should still be able to cancel out of the command early if they choose without terminating the entire `pwsh` process.

I understand that it's not especially simple code, and ideally we would use a Task-based async method with an easily-defined `cancellationToken`. However, currently the .NET Core `Ping.SendPingAsync()` method does not support a `cancellationToken` to enable this. This is the next best thing until that is available from the API.